### PR TITLE
Implement flutter/mousecursor

### PR DESCRIPTION
**Pull request by provokateurin**
Implementation of https://github.com/flutter/flutter/commit/7f8cb7f83038014f3371d085c2722942ef0e6d5a
Cursor list from https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/rendering/mouse_cursor.dart

### CODE REVIEW

File: mousecursor.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+// flutter/mousecursor channel.
+type mousecursorPlugin struct {
+	channel   *plugin.MethodChannel
+	messenger plugin.BinaryMessenger
+	window    *glfw.Window
+}
+
+var defaultMousecursorPlugin = &mousecursorPlugin{}
+
+var _ PluginGLFW = &mousecursorPlugin{} // compile-time type check
```
Comment (pchampio) [inline mousecursor.go:22]: please remove this compile-time check, as #435 we are removing them

File: mousecursor.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+func (p *mousecursorPlugin) InitPlugin(messenger plugin.BinaryMessenger) error {
+	p.messenger = messenger
+
+	return nil
+}
+func (p *mousecursorPlugin) InitPluginGLFW(window *glfw.Window) error {
+	p.window = window
+	p.channel = plugin.NewMethodChannel(p.messenger, mousecursorChannelName, plugin.StandardMethodCodec{})
```
Comment (pchampio) [inline mousecursor.go:31]: define the p.channel in the InitPlugin.

File: mousecursor.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	messenger plugin.BinaryMessenger
+	window    *glfw.Window
+}
+
+var defaultMousecursorPlugin = &mousecursorPlugin{}
+
+var _ PluginGLFW = &mousecursorPlugin{} // compile-time type check
+
+func (p *mousecursorPlugin) InitPlugin(messenger plugin.BinaryMessenger) error {
+	p.messenger = messenger
```
Comment (pchampio) [inline mousecursor.go:25]: remove `p.messenger` from the struct

File: mousecursor.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/go-gl/glfw/v3.3/glfw"
+)
+
+const mousecursorChannelName = "flutter/mousecursor"
+
+// mousecursorPlugin implements flutter.Plugin and handles method calls to the
+// flutter/mousecursor channel.
+type mousecursorPlugin struct {
+	channel   *plugin.MethodChannel
+	messenger plugin.BinaryMessenger
```
Comment (pchampio) [inline mousecursor.go:16]: channel already has the pointer to a `messenger`

File: mousecursor.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+	return nil
+}
+func (p *mousecursorPlugin) InitPluginGLFW(window *glfw.Window) error {
+	p.window = window
+	p.channel = plugin.NewMethodChannel(p.messenger, mousecursorChannelName, plugin.StandardMethodCodec{})
+	p.channel.HandleFuncSync("activateSystemCursor", p.handleActivateSystemCursor)
+	return nil
+}
+
+func (p *mousecursorPlugin) handleActivateSystemCursor(arguments interface{}) (reply interface{}, err error) {
```
Comment (pchampio) [inline mousecursor.go:36]: In the glfw doc there is https://godoc.org/github.com/go-gl/glfw/v3.3/glfw#Cursor.Destroy
Should we use it?

**Review by pchampio — CHANGES_REQUESTED** [review]

File: mousecursor.go (hunk 5 above)
Comment (jld3103) [reply mousecursor.go:36]: Yes

**Comment by pchampio** [comment]
Newer version of glfw have support for a wider range of cursor type.
https://github.com/glfw/glfw/commit/7dbdd2e6a5f01d2a4b377a197618948617517b0e
This patch isn't tagged in v3.3, will have to wait..

File: mousecursor.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/go-flutter-desktop/go-flutter/plugin"
+	"github.com/go-gl/glfw/v3.3/glfw"
+)
+
+const mousecursorChannelName = "flutter/mousecursor"
+
+// mousecursorPlugin implements flutter.Plugin and handles method calls to the
+// flutter/mousecursor channel.
+type mousecursorPlugin struct {
+	channel    *plugin.MethodChannel
```
Comment (pchampio) [inline mousecursor.go:15]: Sorry, didn't catch this in the first review.
p.channel can also be removed from the plugin struct.

File: mousecursor.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+var defaultMousecursorPlugin = &mousecursorPlugin{}
+
+func (p *mousecursorPlugin) InitPlugin(messenger plugin.BinaryMessenger) error {
+	p.channel = plugin.NewMethodChannel(messenger, mousecursorChannelName, plugin.StandardMethodCodec{})
+	return nil
+}
+func (p *mousecursorPlugin) InitPluginGLFW(window *glfw.Window) error {
+	p.window = window
+	p.channel.HandleFuncSync("activateSystemCursor", p.handleActivateSystemCursor)
```
Comment (pchampio) [inline mousecursor.go:28]: The HandleFuncSync func can be moved in to InitPlugin func. Then p.channel can be removed

File: mousecursor.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+type mousecursorPlugin struct {
+	channel    *plugin.MethodChannel
+	window     *glfw.Window
+	lastCursor *glfw.Cursor
+}
+
+var defaultMousecursorPlugin = &mousecursorPlugin{}
+
+func (p *mousecursorPlugin) InitPlugin(messenger plugin.BinaryMessenger) error {
+	p.channel = plugin.NewMethodChannel(messenger, mousecursorChannelName, plugin.StandardMethodCodec{})
```
Comment (pchampio) [inline mousecursor.go:23]: make p.channel a local var

**Review by pchampio — CHANGES_REQUESTED** [review]

**Review by pchampio — APPROVED** [review]

File: mousecursor.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+		cursor = glfw.CreateStandardCursor(glfw.HandCursor)
+	case kind == "text":
+		cursor = glfw.CreateStandardCursor(glfw.IBeamCursor)
+	default:
+		return nil, errors.New(fmt.Sprintf("cursor kind %s not implemented", args["kind"]))
+	}
+	if p.lastCursor != nil {
+		p.lastCursor.Destroy()
+	}
+	if cursor != nil {
```
Comment (GeertJohan) [inline mousecursor.go:55]: I'm concerned about this flow.

What if lastCursor isn't nil and cursor IS nil, then the previous cursor is destroyed and no new one is set?

Looking at the code above, we won't end up here without cursor being assigned, but if that should be a given, then we shouldn't have thi if statement here.

Also wondering: should we set the new cursor before destroying the old one?

File: mousecursor.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	case kind == "basic" || kind == "forbidden" || kind == "grab" || kind == "grabbing":
+		// go-gl GLFW currently (latest tagged v3.3 version) has no cursors for "forbidden", "grab" and "grabbing"
+		// TODO: Wait for https://github.com/glfw/glfw/commit/7dbdd2e6a5f01d2a4b377a197618948617517b0e to appear in go-gl GLFW and implement the "forbidden" cursor
+		cursor = glfw.CreateStandardCursor(glfw.ArrowCursor)
+	case kind == "click":
+		cursor = glfw.CreateStandardCursor(glfw.HandCursor)
+	case kind == "text":
+		cursor = glfw.CreateStandardCursor(glfw.IBeamCursor)
+	default:
+		return nil, errors.New(fmt.Sprintf("cursor kind %s not implemented", args["kind"]))
```
Comment (pchampio) [inline mousecursor.go:50]: mousecursor.go:50:15: should replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)

**Comment by davidmartos96** [comment]
@jld3103 Can this feature be tested in any of the Flutter versions (including master)? It looks interesting to try out at least.

**Comment by jld3103** [comment]
I think it is still not in beta, but you can just try it on master

**Comment by davidmartos96** [comment]
@jld3103 Thanks! It actually works in the current beta branch. It was my bad, when I tried it, I was a bit behind in the go-flutter version.

**Pull request merged**